artiq/test/coredevice/test_moninj.py

```diff
@@ -66,3 +71,237 @@ def injection_status_cb(channel, override, value):
             (loop_out_channel, TTLOverride.en.value, 1),
             (loop_out_channel, TTLOverride.en.value, 1)
         ])
+
+
+class _UrukulExperiment(EnvExperiment):
+    def build(self):
+        self.setattr_device("core")
+        
+    @kernel
+    def init_channel(self, channel):
+        self.core.reset()
+        self.core.break_realtime()
+        channel.cpld.init()
+        channel.init()
+
+    @kernel
+    def read_raw(self, channel):
+        self.init_channel(channel)
+        return channel.get_mu()
+
+    @kernel
+    def write_raw(self, channel, ftw, pow: TInt32 = 0):
+        self.init_channel(channel)
+        channel.set_mu(ftw, pow)
+
+    @kernel
+    def write_raw_with_asf(self, channel, ftw, pow: TInt32 = 0, asf: TInt32 = 0):
+        self.init_channel(channel)
+        channel.set_mu(ftw, pow, asf)
+
+    @kernel
+    def read(self, channel):
+        self.init_channel(channel)
+        return channel.get()
+
+    @kernel
+    def write(self, channel, freq: TFloat, phase: TFloat = 0.0):
+        self.init_channel(channel)
+        channel.set(freq, phase)
+
+    @kernel
+    def write_with_amp(self, channel, freq: TFloat, phase: TFloat = 0.0, amplitude: TFloat = 0.0):
+        self.init_channel(channel)
+        channel.set(freq, phase, amplitude)
+
+
+class AD991XMonitorTest(ExperimentCase, IsolatedAsyncioTestCase):
+    def get_urukuls(self):
+        urukuls = defaultdict(dict)
+        ddb = self.device_db.get_device_db()
+        for name, desc in ddb.items():
+            if isinstance(desc, dict) and desc["type"] == "local":
+                module, cls = desc["module"], desc["class"]
+                if (module, cls) == ("artiq.coredevice.ad9910", "AD9910"):
+                    urukuls[cls][name] = self.device_mgr.get(name)
+                if (module, cls) == ("artiq.coredevice.ad9912", "AD9912"):
+                    urukuls[cls][name] = self.device_mgr.get(name)
+        return urukuls
+
+    def urukuls_all(self):
+        ret = {}
+        for x in self.urukuls.values():
+            ret.update(x)
+        return ret
+
+    def ensure_ad9910_only(self):
```

Comment:
Poorly named function.

Reply:
this is now removed since i separated the test cases into two different class which implies only this kind of device would run this test.